File: wpf-dev-pack/hooks/xaml-validator.js
#!/usr/bin/env node

/**
 * XAML Validator Hook
 * XAML 검증 Hook
 *
 * Validates XAML files after Write/Edit operations for common issues.
 * Write/Edit 작업 후 XAML 파일의 일반적인 문제를 검증합니다.
 */

const fs = require('fs');
const path = require('path');

// Common XAML issues to check
// 검사할 일반적인 XAML 문제들
const validationRules = [
  {
    name: 'Missing x:Key in Style without TargetType',
    pattern: /<Style\s+(?!.*x:Key)(?!.*TargetType)/i,
    message: 'Style should have either x:Key or TargetType attribute'
  },
  {
    name: 'Direct content in Generic.xaml',
    pattern: /<Style\s+.*TargetType.*(?<!MergedDictionaries)/i,
    checkFile: (filePath) => filePath.toLowerCase().includes('generic.xaml'),
    message: 'Generic.xaml should only contain MergedDictionaries, not direct styles'
  },
  {
    name: 'Missing xmlns:x namespace',
    pattern: /<ResourceDictionary(?![\s\S]*xmlns:x=)/i,
    message: 'ResourceDictionary should include xmlns:x namespace declaration'
  },
  {
    name: 'TemplateBinding outside ControlTemplate',
    pattern: /<(?!ControlTemplate)[\w.]+[^>]*TemplateBinding/i,
    message: 'TemplateBinding should only be used inside ControlTemplate'
  },
  {
    name: 'Missing PART_ prefix for template parts',
    pattern: /x:Name="(?!PART_)[A-Z][^"]*".*(?:ContentPresenter|Border|Grid)/i,
    message: 'Template parts should use PART_ prefix (e.g., PART_ContentHost)'
  }
];

function validateXaml(content, filePath) {
  const issues = [];

  for (const rule of validationRules) {
    // Skip if rule has file-specific check that doesn't match
    // 파일별 검사가 있고 매치되지 않으면 건너뜀
    if (rule.checkFile && !rule.checkFile(filePath)) {
      continue;
    }

    if (rule.pattern.test(content)) {
      issues.push({
        rule: rule.name,
        message: rule.message
      });
    }
  }

  return issues;
}

function main() {
  try {
    // Get file path from environment
    // 환경 변수에서 파일 경로 가져오기
    const filePath = process.env.CLAUDE_TOOL_FILE_PATH || '';

    if (!filePath) {
      process.exit(0);
    }

    // Only validate XAML files
    // XAML 파일만 검증
    const ext = path.extname(filePath).toLowerCase();
    if (ext !== '.xaml') {
      process.exit(0);
    }

    // Check if file exists
    // 파일 존재 여부 확인
    if (!fs.existsSync(filePath)) {
      process.exit(0);
    }

    const content = fs.readFileSync(filePath, 'utf-8');
    const issues = validateXaml(content, filePath);

    if (issues.length > 0) {
      console.log(`[WPF Dev Pack] XAML validation warnings for ${path.basename(filePath)}:`);
      issues.forEach(issue => {
        console.log(`  - ${issue.rule}: ${issue.message}`);
      });
    }

    process.exit(0);
  } catch (error) {
    console.error(`[WPF Dev Pack] Validation error: ${error.message}`);
    process.exit(0); // Don't block on error
  }
}

main();

File: wpf-dev-pack/hooks/XamlValidator.cs
#!/usr/bin/env dotnet

// XAML Validator Hook
// Validates XAML files after Write/Edit operations for common issues.
// Input: stdin JSON with "tool_name" and "tool_input.file_path" fields

using System.Text.Json;
using System.Text.RegularExpressions;

// Read JSON from stdin
var input = Console.In.ReadToEnd();
if (string.IsNullOrWhiteSpace(input))
    return;

string? toolName = null;
string? filePath = null;

try
{
    using var doc = JsonDocument.Parse(input);

    if (doc.RootElement.TryGetProperty("tool_name", out var tn))
        toolName = tn.GetString();

    if (doc.RootElement.TryGetProperty("tool_input", out var ti))
    {
        if (ti.TryGetProperty("file_path", out var fp))
            filePath = fp.GetString();
    }
}
catch
{
    return;
}

if (toolName is not "Write" and not "Edit")
    return;

if (string.IsNullOrEmpty(filePath))
    return;

var ext = Path.GetExtension(filePath).ToLowerInvariant();
if (ext != ".xaml")
    return;

if (!File.Exists(filePath))
    return;

var content = File.ReadAllText(filePath);
var issues = XamlValidationRules.Validate(content, filePath);

if (issues.Count > 0)
{
    Console.WriteLine($"[WPF Dev Pack] XAML validation warnings for {Path.GetFileName(filePath)}:");
    foreach (var issue in issues)
        Console.WriteLine($"  - {issue.Rule}: {issue.Message}");
}

/// <summary>
/// XAML validation rules using GeneratedRegex for optimal performance.
/// </summary>
internal static partial class XamlValidationRules
{
    public static List<(string Rule, string Message)> Validate(string content, string filePath)
    {
        var issues = new List<(string Rule, string Message)>();

        // Rule 1: Missing x:Key in Style without TargetType
        if (StyleWithoutKeyOrTargetType().IsMatch(content))
            issues.Add(("Missing x:Key in Style without TargetType", "Style should have either x:Key or TargetType attribute"));

        // Rule 2: Direct content in Generic.xaml
        if (filePath.Contains("generic.xaml", StringComparison.OrdinalIgnoreCase) &&
            StyleWithTargetType().IsMatch(content))
            issues.Add(("Direct content in Generic.xaml", "Generic.xaml should only contain MergedDictionaries, not direct styles"));

        // Rule 3: Missing xmlns:x namespace
        if (ResourceDictionaryWithoutXmlnsX().IsMatch(content))
            issues.Add(("Missing xmlns:x namespace", "ResourceDictionary should include xmlns:x namespace declaration"));

        // Rule 4: TemplateBinding outside ControlTemplate
        if (TemplateBindingOutsideControlTemplate().IsMatch(content))
            issues.Add(("TemplateBinding outside ControlTemplate", "TemplateBinding should only be used inside ControlTemplate"));

        // Rule 5: Missing PART_ prefix for template parts
        if (MissingPartPrefix().IsMatch(content))
            issues.Add(("Missing PART_ prefix for template parts", "Template parts should use PART_ prefix (e.g., PART_ContentHost)"));

        return issues;
    }

    [GeneratedRegex(@"<Style\s+(?!.*x:Key)(?!.*TargetType)", RegexOptions.IgnoreCase)]
    private static partial Regex StyleWithoutKeyOrTargetType();

    [GeneratedRegex(@"<Style\s+.*TargetType.*(?<!MergedDictionaries)", RegexOptions.IgnoreCase)]
    private static partial Regex StyleWithTargetType();

    [GeneratedRegex(@"<ResourceDictionary(?![\s\S]*xmlns:x=)", RegexOptions.IgnoreCase)]
    private static partial Regex ResourceDictionaryWithoutXmlnsX();

    [GeneratedRegex(@"<(?!ControlTemplate)[\w.]+[^>]*TemplateBinding", RegexOptions.IgnoreCase)]
    private static partial Regex TemplateBindingOutsideControlTemplate();

    [GeneratedRegex(@"x:Name=""(?!PART_)[A-Z][^""]*"".*(?:ContentPresenter|Border|Grid)", RegexOptions.IgnoreCase)]
    private static partial Regex MissingPartPrefix();
}
